tests/common/util.rs

```diff
@@ -1191,26 +1221,613 @@ impl UCommand {
+            None if self.raw.stderr.is_some() => {
+                let stderr = self.raw.stderr.as_mut().unwrap();
+                let mut buffer = vec![0; size];
+                stderr.read_exact(&mut buffer).unwrap();
+                buffer
+            }
+            None => vec![],
+        }
+    }
+
+    /// Pipe data into [`Child`] stdin in a separate thread to avoid deadlocks. In contrast to
+    /// [`UChild::write_in`], this method is designed to simulate a pipe on the command line and can
+    /// be used only once or else panics. Note, that [`UCommand::set_stdin`] must be used together
+    /// with [`Stdio::piped`] or else this method doesn't work as expected. `Stdio::piped` is the
+    /// current default when using [`UCommand::run_no_wait`]) without calling `set_stdin`. This
+    /// method stores a [`JoinHandle`] of the thread in which the writing to the child processes'
+    /// stdin is running. The associated thread is joined with the main process in the methods below
+    /// when exiting the child process.
+    ///
+    /// * [`UChild::wait`]
+    /// * [`UChild::wait_with_output`]
+    /// * [`UChild::pipe_in_and_wait`]
+    /// * [`UChild::pipe_in_and_wait_with_output`]
+    ///
+    /// Usually, there's no need to join manually but if needed, the [`UChild::join`] method can be
+    /// used .
+    ///
+    /// [`JoinHandle`]: std::thread::JoinHandle
+    pub fn pipe_in<T: Into<Vec<u8>>>(&mut self, content: T) -> &mut Self {
+        let ignore_stdin_write_error = self.ignore_stdin_write_error;
+        let content = content.into();
+        let stdin = self
+            .raw
+            .stdin
+            .take()
+            .expect("Could not pipe into child process. Was it set to Stdio::null()?");
+
+        let join_handle = thread::spawn(move || {
+            let mut writer = BufWriter::new(stdin);
+
+            match writer.write_all(&content).and_then(|_| writer.flush()) {
+                Err(error) if !ignore_stdin_write_error => Err(io::Error::new(
+                    io::ErrorKind::Other,
+                    format!("failed to write to stdin of child: {}", error),
+                )),
+                Ok(_) | Err(_) => Ok(()),
+            }
+        });
+
+        self.join_handle = Some(join_handle);
+        self
+    }
+
+    /// Call join on the thread created by [`UChild::pipe_in`] and if the thread is still running.
+    /// This method can be called multiple times but is a noop if already joined.
+    pub fn join_in(&mut self) -> &mut Self {
+        if let Some(join_handle) = self.join_handle.take() {
+            join_handle
+                .join()
+                .expect("Error joining with the piping stdin thread")
+                .unwrap();
+        }
+        self
+    }
+
+    /// Convenience method for [`UChild::pipe_in`] and then [`UChild::wait`]
+    pub fn pipe_in_and_wait<T: Into<Vec<u8>>>(mut self, content: T) -> CmdResult {
```

Comment:
Same as `make_assertion_with_delay`. I guess it's personal preference, but I'd like to keep the API small, so I think these are not necessary. I do like the `try_` methods though, because they fit with the rest of the ecosystem.

Reply:
I really like having convenience methods for common usage patterns. This one saves you writing two lines instead of just one. The `wait()` methods consumes `UChild`, so it's not possible to write `self.pipe_in("something").wait().unwrap()` without getting a compiler error.

Reply:
Oh that's annoying. I agree then, though it is something I'd like to clean up later by making `pipe_in` take ownership of `self`, but that would be a big breaking change.